src/Microsoft.SqlTools.ServiceLayer/DacFx/DacFxOperation.cs

```diff
@@ -0,0 +1,122 @@
+using Microsoft.SqlTools.ServiceLayer.TaskServices;
+using Microsoft.SqlTools.Utility;
+using System;
+using System.Data.SqlClient;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.SqlTools.ServiceLayer.DacFx
+{
+    /// <summary>
+    /// Base class for DacFx operations
+    /// </summary>
+    abstract class DacFxOperation : ITaskOperation
+    {
+        private CancellationTokenSource cancellation = new CancellationTokenSource();
+        private bool disposed = false;
+
+        /// <summary>
+        /// Gets the unique id associated with this instance.
+        /// </summary>
+        public string OperationId { get; private set; }
+
+        public SqlTask SqlTask { get; set; }
+
+        protected SqlConnection sqlConnection { get; set; }
+
+        protected DacServices dacServices { get; set; }
+
+        protected DacFxOperation(SqlConnection sqlConnection)
+        {
+            Validate.IsNotNull("sqlConnection", sqlConnection);
+            this.sqlConnection = sqlConnection;
+            this.OperationId = Guid.NewGuid().ToString();
+        }
+
+        protected CancellationToken CancellationToken { get { return this.cancellation.Token; } }
+
+        /// <summary>
+        /// The error occurred during operation
+        /// </summary>
+        public string ErrorMessage
+        {
+            get
+            {
+                return string.Empty;
+            }
+        }
+
+        /// <summary>
+        /// Cancel operation
+        /// </summary>
+        public void Cancel()
+        {
+            if (!this.cancellation.IsCancellationRequested)
+            {
+                Logger.Write(TraceEventType.Verbose, string.Format("Cancel invoked for OperationId {0}", this.OperationId));
+                this.cancellation.Cancel();
+            }
+        }
+
+        /// <summary>
+        /// Disposes the operation.
+        /// </summary>
+        public void Dispose()
+        {
+            if (!disposed)
+            {
+                this.Cancel();
+                disposed = true;
+            }
+        }
+
+        public void Execute(TaskExecutionMode mode)
+        {
+            if (this.CancellationToken.IsCancellationRequested)
+            {
+                throw new OperationCanceledException(this.CancellationToken);
+            }
+
+            try
+            {
+                this.dacServices = new DacServices(this.sqlConnection.ConnectionString);
+                this.dacServices.ProgressChanged += new EventHandler<DacProgressEventArgs>(this.DacProgressUpdates);
+                Execute();
+            }
+            catch (Exception e)
+            {
+                Logger.Write(TraceEventType.Error, string.Format("DacFx import operation {0} failed with exception {1}", this.OperationId, e));
+                throw;
+            }
+        }
+
+        public abstract void Execute();
+
+        protected void DacProgressUpdates(object obj, DacProgressEventArgs args)
```

Comment:
We're unbinding event handlers - I generally don't expect in an event handler to unbind themselves, but to do something with the event arg data. What is the method attempting to accomplish?

Reply:
Removed for now since the event handler isn't doing anything